### ISSUE TITLE
Can't type `d` character and Helix crash

### ISSUE THREAD

**Issue by d4hines** [helix-editor/helix#5613]
### Summary

When trying to type the following code, I noticed that I could not type the `d` character in `payload`:
```
 let signature = String.sub payload
 ```
 Pressing tab to complete the variable works. But then when trying to finish the line like so:

```
 let signature = String.sub payload 0 99 |> Ed25519...
```
I can't press the 'd' in 'Ed25519' either.

When switching away from helix after reproducing this error, Helix crashed twice. I couldn't reproduce the crashing behavior after that, so I'm not sure if they're related.

### Reproduction Steps

 

I tried this:
1. Open this file: https://gitlab.com/marigold/tezos/-/blob/d4hines/hacking-on-rollups/src/proto_alpha/lib_protocol/sc_rollup_arith.ml
2. I'm using the nix shell provided in the root of the repository. It provides `ocaml-lsp`.
3. Go to line 984.
4. Try to insert a new line beneath with `o`.
5. Try to type `let signature = String.sub payload`

I expected this to happen:
The character `d` should be inserted.

Instead, this happened:
Nothing happens.

### Helix log

<details><summary>~/.cache/helix/helix.log</summary>

```
2023-01-20T16:45:26.568 helix_lsp::transport [ERROR] err <- "/-----------------------------------------------------------------------\n"
2023-01-20T16:45:26.568 helix_lsp::transport [ERROR] err <- "| Internal error: Uncaught exception.\n"
2023-01-20T16:45:26.568 helix_lsp::transport [ERROR] err <- "| Jsonrpc: json conversion failed: missing field error\n"
2023-01-20T16:45:26.569 helix_lsp::transport [ERROR] err <- "| Raised at Jsonrpc__Import.Json.error in file \"jsonrpc/src/import.ml\", line 29, characters 23-50\n"
2023-01-20T16:45:26.569 helix_lsp::transport [ERROR] err <- "| Called from Jsonrpc.Response.t_of_yojson in file \"jsonrpc/src/jsonrpc.ml\", line 244, characters 18-74\n"
2023-01-20T16:45:26.569 helix_lsp::transport [ERROR] err <- "| Called from Lsp__Io.Make.read.resp in file \"lsp/src/io.ml\", line 99, characters 37-72\n"
2023-01-20T16:45:26.569 helix_lsp::transport [ERROR] err <- "| Called from Lsp__Io.Make.read in file \"lsp/src/io.ml\" (inlined), line 102, characters 6-18\n"
2023-01-20T16:45:26.569 helix_lsp::transport [ERROR] err <- "| Called from Lsp__Io.Make.read.(fun) in file \"lsp/src/io.ml\", line 115, characters 26-39\n"
2023-01-20T16:45:26.569 helix_lsp::transport [ERROR] err <- "| Called from Fiber__Scheduler.exec in file \"src/fiber/scheduler.ml\", line 69, characters 8-11\n"
2023-01-20T16:45:26.569 helix_lsp::transport [ERROR] err <- "| Re-raised at Stdune__Exn.raise_with_backtrace in file \"otherlibs/stdune/exn.ml\" (inlined), line 36, characters 27-56\n"
2023-01-20T16:45:26.569 helix_lsp::transport [ERROR] err <- "| Called from Stdune__Exn_with_backtrace.reraise in file \"otherlibs/stdune/exn_with_backtrace.ml\", line 18, characters 33-71\n"
2023-01-20T16:45:26.569 helix_lsp::transport [ERROR] err <- "| Called from Fiber__Scheduler.advance.(fun) in file \"src/fiber/scheduler.ml\", line 191, characters 2-58\n"
2023-01-20T16:45:26.569 helix_lsp::transport [ERROR] err <- "| Called from Fiber.run.loop in file \"src/fiber/fiber.ml\", line 15, characters 30-61\n"
2023-01-20T16:45:26.569 helix_lsp::transport [ERROR] err <- "| Called from Stdune__Exn_with_backtrace.try_with in file \"otherlibs/stdune/exn_with_backtrace.ml\", line 9, characters 8-12\n"
2023-01-20T16:45:26.569 helix_lsp::transport [ERROR] err <- "\\-----------------------------------------------------------------------\n"
2023-01-20T16:45:26.569 helix_lsp::transport [ERROR] err <- "\n"
2023-01-20T16:45:26.586 helix_lsp::client [WARN] language server failed to terminate gracefully - server closed the stream
2023-01-20T16:45:26.586 helix_lsp::transport [ERROR] err: <- IO(Os { code: 32, kind: BrokenPipe, message: "Broken pipe" })
2023-01-20T16:45:26.589 helix_lsp::transport [ERROR] err: <- StreamClosed
```

</details>

### Platform

Nixos

### Terminal Emulator

kitty 0.26.2

### Helix Version

22.12-102-g0bce1fd1

**Comment by d4hines** [comment]
An important detail I forgot in the report is that I'm rebased on top of https://github.com/helix-editor/helix/pull/2377 and, more relevantly, https://github.com/helix-editor/helix/pull/5420. I'll try testing against master later.

**Comment by CBenoit** [comment]
I believe it will be a bit hard to investigate without the panic stack trace. Next time this happens, do you think you could copy and paste it here? I recommend enabling  full backtrace with env variable `RUST_BACKTRACE=1` when testing on the bleeding edge, just in case. Also, if you’re able to reproduce, could you try again against this branch?
https://github.com/pascalkuthe/helix/compare/render_system_2...CBenoit:render_system_2_cbenoit_fork?expand=1
Thank you!

**Comment by the-mikedavis** [comment]
I can't reproduce this running from master or #5420 using the Nix shell. I spawn the Nix shell with `nix-shell` and run `dune build` within. Then the reproduction steps as listed also within the shell.

Also, could you change the source link to a permalink? The line number could change if you push to that branch and make this harder to reproduce.

**Comment by pascalkuthe** [comment]
> An important detail I forgot in the report is that I'm rebased on top of #2377 and, more relevantly, #5420. I'll try testing against master later.

I will keep this issue open so the discussion doesn't get lost but if you encounter bugs while running #5420 please report them as a comment against that PR instead of as a new issue. Issues are only intended for bug reports in current master.

@CBenoit reported a similar bug there. I will try to dig into the cause. One major problem for me is that I am unable to reproduce the crash (despite the great instructions provided by @CBenoit) so I am still a bit in the dark about what is actually going on

**Comment by d4hines** [comment]
I believe I figured out the issue and it was my mistake - in my branch, I had done a remapping of the character `d` in popups while hunting a performance issue. When the completion popup was triggered, I got this behavior.

As to the crash, I wasn't able to reproduce it consistently. In any case I've rebased on top of the latest changes to #5420. I'll set `RUST_BACKTRACE=1` on my system and see if it happens again.
